### PR TITLE
Replace AssessmentCategory model with assessment_name field

### PR DESCRIPTION
**Pull request by icanberk**

### CODE REVIEW

**Comment by icanberk** [comment]
I'm not sure if this was worth it, but I've done it already.

This needs to be tested with a large enough data set since it touches gradebook.

File: app/models/assessment.rb [hunk 1]
```diff
@@ -30,8 +29,9 @@ class Assessment < ActiveRecord::Base
                             :greater_than_or_equal_to => -1, :allow_nil => true
   validates_numericality_of :max_grace_days, :only_integer => true,
                             :greater_than_or_equal_to => 0, :allow_nil => true
-  validates_presence_of :name, :display_name, :due_at, :end_at, :start_at, :quiz, :quizData,
-                        :grading_deadline, :category_id, :max_size, :max_submissions
+
+  validates_presence_of :name, :display_name, :due_at, :end_at, :start_at,
+                        :grading_deadline, :max_size, :max_submissions
```
Comment (dlbucci) [inline app/models/assessment.rb]: Don't we still want each assessment to have a category name?

File: app/models/course.rb [hunk 2]
```diff
@@ -142,6 +141,13 @@ def students
     course_user_data.where(course_assistant: false, instructor: false, dropped: [false, nil])
   end
 
+  def assessment_categories
+    assessments.pluck(:category_name).uniq.sort
+  end
+
+  def assessments_with_category(cat_name)
```
Comment (dlbucci) [inline app/models/course.rb]: I think assessments_in_category or assessments_by_category might be more descriptive names, but it's your call.

File: app/models/course.rb [hunk 3]
```diff
@@ -142,6 +141,13 @@ def students
     course_user_data.where(course_assistant: false, instructor: false, dropped: [false, nil])
   end
 
+  def assessment_categories
+    assessments.pluck(:category_name).uniq.sort
```
Comment (dlbucci) [inline app/models/course.rb]: uniq is an alias for distinct, which seems to be the newer, cooler way to specify this, so maybe go with that?  No big deal.

File: db/migrate/20150130054629_add_category_to_assessments.rb [hunk 4]
```diff
@@ -0,0 +1,18 @@
...
+    
+    Assessment.all.each do |assessment|
+        sql = "SELECT name FROM assessment_categories WHERE id=" + assessment.category_id.to_s
+		records_array = ActiveRecord::Base.connection.execute(sql)
+	   	
+	   	records_array.each do |cat|
+    		assessment.category_name = cat[0]
+			assessment.save
+	    end
+    end
```
Comment (dlbucci) [inline db/migrate/20150130054629_add_category_to_assessments.rb]: No big deal, but indentation is off here.  What's wrong with adding the category name column, then running something like
```
Assessment.all.each do |assessment|
  assessment.category_name = assessment.category.name
  assessment.save!
end
```

File: db/schema.rb [hunk 5]
```diff
@@ -90,9 +87,10 @@
     t.boolean  "has_partners",       limit: 1
     t.boolean  "has_scoreboard",     limit: 1
     t.boolean  "has_svn",            limit: 1
-    t.string   "remote_handin_path", limit: 255
     t.boolean  "quiz",               limit: 1,     default: false
     t.text     "quizData",           limit: 65535
+    t.string   "remote_handin_path", limit: 255
+    t.string   "category_name",      limit: 255
```
Comment (dlbucci) [inline db/schema.rb]: Would we maybe want this to be just 'category'?  the '_name' seems unnecessary, but it's no big deal.

Comment (icanberk) [inline db/schema.rb]: I think it's good to be more explicit. (also makes it easy to refactor/debug for me)

File: db/migrate/20150130054629_add_category_to_assessments.rb (hunk 4 above)
Comment (icanberk) [inline db/migrate/20150130054629_add_category_to_assessments.rb]: Problem was that we removed AssessmentCategory from our codebase, so `category.name` does not work.

I'll fix the indentation.

Comment (dlbucci) [inline db/migrate/20150130054629_add_category_to_assessments.rb]: Ah, well this should be fine, but yeah, I'm not sure we want to drop AssessmentCategories so quickly.  We may want to keep that table around for a bit until we're sure that grades aren't getting screwed up...

File: app/models/assessment.rb (hunk 1 above)
Comment (icanberk) [inline app/models/assessment.rb]: Yeah, I think we need to force this since a lot of things like category average depends on this.

**Comment by icanberk** [comment]
I've tested this with real data on sawshark. Everything seems to work well. I've also almost documented a way of (manually) testing gradebook changes.

File: app/models/assessment.rb (hunk 1 above)
Comment (dlbucci) [inline app/models/assessment.rb]: Right, so add it back to the validates_presence_of?

Comment (icanberk) [inline app/models/assessment.rb]: My bad! Fixed.

File: db/schema.rb [hunk 6]
```diff
@@ -40,15 +40,12 @@
...
-    t.string  "name",      limit: 255
-    t.integer "course_id", limit: 4
-  end
-
   create_table "assessment_user_data", force: :cascade do |t|
     t.integer  "course_user_datum_id", limit: 4,               null: false
     t.integer  "assessment_id",        limit: 4,               null: false
     t.integer  "latest_submission_id", limit: 4
+    t.integer  "cgdub",                limit: 4
+    t.integer  "special_type",         limit: 4,   default: 0, null: false
```
Comment (dlbucci) [inline db/schema.rb]: Where are these columns coming from?

File: app/models/grade_matrix.rb [hunk 7]
```diff
@@ -103,8 +102,8 @@ def matrix!
 
       @course.assessment_categories.each do |cat|
         a = cud.category_average(cat, @as_seen_by)
-        cat_avg_by_cat["#{cat.id}"] ||= {}
-        cat_avg_by_cat["#{cat.id}"]["#{cud.id}"] = a
+        cat_avg_by_cat["#{cat}"] ||= {}
+        cat_avg_by_cat["#{cat}"]["#{cud.id}"] = a
```
Comment (dlbucci) [inline app/models/grade_matrix.rb]: This could just be cat_avg_by_cat[cat].

File: db/schema.rb (hunk 6 above)
Comment (icanberk) [inline db/schema.rb]: No idea...

Comment (dlbucci) [inline db/schema.rb]: Can we figure that out and possibly get rid of these if we don't need them?

Comment (icanberk) [inline db/schema.rb]: Yeah, but I think we should do it in another branch

Comment (icanberk) [inline db/schema.rb]: Yeah, but I think we should do it in another branch. I'll try to cleanup for now.

File: db/migrate/20150203151605_cleanup_old_fields.rb [hunk 8]
```diff
@@ -22,9 +22,17 @@ def change
...
+        remove_column :submissions, :tweak_old, :float
+    end
+    
+    if column_exists? :submissions, :ignored_old
+        remove_column :submissions, :ignored_old, :boolean
+    end
+
+    if column_exists? :submissions, :absolute_tweak  
+        remove_column :submissions, :absolute_tweak, :boolean
+    end
```
Comment (dlbucci) [inline db/migrate/20150203151605_cleanup_old_fields.rb]: Has this migration already been ran in production?  If so, we should just make a new migration.

Comment (icanberk) [inline db/migrate/20150203151605_cleanup_old_fields.rb]: Nope. I just ran it on staging.

Comment (icanberk) [inline db/migrate/20150203151605_cleanup_old_fields.rb]: Yeah it did... Do we really need to? It essentially does the same thing.

Comment (dlbucci) [inline db/migrate/20150203151605_cleanup_old_fields.rb]: It won't run if you change an old migration file that's been run.  You'd need to rollback, which introduces a bunch of potential problems.

Comment (icanberk) [inline db/migrate/20150203151605_cleanup_old_fields.rb]: I don't want it to run. Those columns are already removed.

Comment (dlbucci) [inline db/migrate/20150203151605_cleanup_old_fields.rb]: I see.  Why did you need to make this change at all then?

Comment (icanberk) [inline db/migrate/20150203151605_cleanup_old_fields.rb]: I ran migrations on sawshark and this one was giving an error. Not sure
why. It's not a big deal - I can just revert it but I thought the changes
just make it safer.

Comment (dlbucci) [inline db/migrate/20150203151605_cleanup_old_fields.rb]: I just want to be very sure that these migrations are going to run and that the schema file will be accurate.

Comment (icanberk) [inline db/migrate/20150203151605_cleanup_old_fields.rb]: I ran them couple times on staging server. We will be fine.

**Comment by icanberk** [comment]
Can we merge this?

**Comment by dlbucci** [comment]
Could you merge develop back into this so we can automerge?  I'd like to be able to reset my DB with develop HEAD schema, run the migrations in this branch and check that they work on my own machine.

**Comment by icanberk** [comment]
@dlbucci done

**Comment by dlbucci** [comment]
Second note: @icanberk, I'll probably hate you for getting this in before groups, cause now I have to resolve the conflicts!

**Comment by icanberk** [comment]
:love_letter:

**Comment by icanberk** [comment]
Closes #42
